src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java

```diff
@@ -114,11 +165,77 @@ public void write(StreamSession session, DataOutputStreamPlus out, int version)
         CassandraStreamHeader.serializer.serialize(header, out, version);
         out.flush();
 
-        CassandraStreamWriter writer = header.compressionInfo == null ?
-                                       new CassandraStreamWriter(sstable, header.sections, session) :
-                                       new CompressedCassandraStreamWriter(sstable, header.sections,
-                                                                           header.compressionInfo, session);
-        writer.write(out);
+        if (shouldStreamEntireSSTable() && out instanceof ByteBufDataOutputStreamPlus)
+        {
+            CassandraBlockStreamWriter writer = new CassandraBlockStreamWriter(sstable, session, manifest);
+            writer.write((ByteBufDataOutputStreamPlus) out);
+        }
+        else
+        {
+            CassandraStreamWriter writer = (header.compressionInfo == null) ?
+                     new CassandraStreamWriter(sstable, header.sections, session) :
+                     new CompressedCassandraStreamWriter(sstable, header.sections,
+                                                         header.compressionInfo, session);
+            writer.write(out);
+        }
+    }
+
+    @VisibleForTesting
+    public boolean shouldStreamEntireSSTable()
+    {
+        // don't stream if full sstable transfers are disabled or legacy counter shards are present
+        if (!isZeroCopySSTableTransfersEnabled || ref.get().getSSTableMetadata().hasLegacyCounterShards)
+            return false;
+
+        ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(getTableId());
+
+        if (cfs == null)
+            return false;
+
+        AbstractCompactionStrategy compactionStrategy = cfs.getCompactionStrategyManager()
+                                                           .getCompactionStrategyFor(ref.get());
+
+        if (compactionStrategy instanceof LeveledCompactionStrategy)
+            return contained(ranges, ref.get());
+
+        if (compactionStrategy instanceof SizeTieredCompactionStrategy)
+        {
+            return (ranges != null
+                        && ranges.size() == 1
+                        && ranges.get(0)
+                                 .contains(new Range<>(ref.get().first.getToken(),
```

Comment:
Should be Bounds for the sstable, since range is (first, last] and sstable contains [first, last] (thanks @krummas). But on a higher level note, maybe just stick to LCS only for now, and enable STCS after we commit CASSANDRA-14568? We are going to do that in 4.0, so this temporary special case won't make it to a release anyway ¯\_(ツ)_/¯